controllers/tc000243_remediation_retry_test.go

```diff
@@ -0,0 +1,210 @@
+	defer waitResourceToBeDelete(g, &testRepo)
+
+	Given("a Terraform object with auto approve, and attaching it to the bad GitRepository resource.")
+	By("creating a new TF resource and attaching to the bad repo via `sourceRef`.")
+	var helloWorldTF infrav1.Terraform
+	err := helloWorldTF.FromBytes([]byte(fmt.Sprintf(`
+apiVersion: infra.contrib.fluxcd.io/v1alpha2
+kind: Terraform
+metadata:
+  name: %s
+  namespace: flux-system
+spec:
+  approvePlan: auto
+  path: ./terraform-hello-world-example
+  remediation:
+    retries: 3
+  retryInterval: 5s
+  sourceRef:
+    kind: GitRepository
+    name: %s
+    namespace: flux-system
+  interval: 10s
+`, terraformName, sourceName)), runnerServer.Scheme)
+	g.Expect(err).ToNot(HaveOccurred())
+
+	It("should be created and attached successfully.")
+	g.Expect(k8sClient.Create(ctx, &helloWorldTF)).Should(Succeed())
+	defer waitResourceToBeDelete(g, &helloWorldTF)
+
+	By("checking that the TF resource existed inside the cluster.")
+	helloWorldTFKey := types.NamespacedName{Namespace: "flux-system", Name: terraformName}
+	createdHelloWorldTF := infrav1.Terraform{}
+	g.Eventually(func() bool {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		if err != nil {
+			return false
+		}
+		return true
+	}, timeout, interval).Should(BeTrue())
+
+	It("should be reconciled and contain some status conditions.")
+	By("checking that the TF resource's status conditions has 1 element.")
+	g.Eventually(func() int {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		if err != nil {
+			return -1
+		}
+		return len(createdHelloWorldTF.Status.Conditions)
+	}, timeout, interval).Should(Equal(1))
+
+	It("should be an error")
+	By("checking that the Ready's reason of the TF resource become `ArtifactFailed`.")
+	type failedStatusCheckResult struct {
+		Type    string
+		Reason  string
+		Message string
+		Status  metav1.ConditionStatus
+		Retries int64
+	}
+	expected := failedStatusCheckResult{
+		Type:    "Ready",
+		Reason:  "ArtifactFailed",
+		Message: "rpc error: code = Unknown desc = failed to untar artifact, error: requires gzip-compressed body: gzip: invalid header",
+		Status:  metav1.ConditionFalse,
+		Retries: 3,
+	}
+	g.Eventually(func() interface{} {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		if err != nil {
+			return nil
+		}
+		for _, c := range createdHelloWorldTF.Status.Conditions {
+			if c.Type == "Ready" {
+				return failedStatusCheckResult{
+					Type:    c.Type,
+					Reason:  c.Reason,
+					Message: c.Message,
+					Status:  c.Status,
+					Retries: createdHelloWorldTF.Status.ReconciliationFailures,
+				}
+			}
+		}
+		return createdHelloWorldTF.Status
+	}, timeout, interval).Should(Equal(expected))
+
+	// It should never reach 4 when Retry is set to 3.
+	expected = failedStatusCheckResult{
+		Retries: 4,
+	}
+	time.Sleep(15 * time.Second)
+	g.Eventually(func() interface{} {
```

Comment:
If the code is wrong, it could easily be wrong in such a way that it will keep retrying a few times over fifteen seconds, and the count will be higher than 4 the next time it's examined. In other words, this test is likely to pass when you want it to fail. Secondarily: this test will always take more than fifteen seconds to run.

The property under test here is that the system is marked as being in a failed state _if and only if_ the number of failures equals `.Spec.Remediation.Retries`. One direction is demonstrated above: failed state -> number of retries. The other direction is number of retries -> failed state. To show this, you could run the reconciler after it reaches the failure state*, and check that it exits without retrying. These aren't watertight proofs, by any stretch, but they are good enough.

*Since you want to run it _at least_ once, you can just call `r.Reconcile(...)`.

Reply:
Originally I just made a simple `Get` call, but I tend to put **everything** (that contacts kube api) in an `Eventually` block in case the `Get` call returns with an error and causing an intermittent/flaky test failure.

I can see it can pass the check in a faulty situation.

How about this:
```

recheckHelloWorldTF := infrav1.Terraform{}
g.Eventually(func() interface{} {
		return k8sClient.Get(ctx, helloWorldTFKey, &recheckHelloWorldTF)
	}, timeout, interval).Should(gomega.Succeed()))
```
and than check if `recheckHelloWorldTF` and `createdHelloWorldTF` has the same properties/values.

I really don't want to remove that `Eventually`, we had to fix a lot of missing `Eventually` calls just because the API didn't like the result of a d20 roll of the CI at that particular time of the day.

Reply:
> the `Get` call returns with an error and causing an intermittent/flaky test failure

Oh no, an error that's not `NotFound`? What a pain. Since `Eventually` is part of the assertion being made, I think it would be clearer to paper over the unreliable API by using https://pkg.go.dev/k8s.io/client-go/util/retry#OnError, wrapped in a helper. wdyt?

Reply:
I don't really see the difference is a test, the `retry.OnError` would be even more robust just to make sure the test env doesn't :hankey: itself because of low resources:

**With `Eventually`:**
```go
g.Eventually(func() error {
	return k8sClient.Get(ctx, helloWorldTFKey, &recheckHelloWorldTF)
}, timeout, interval)..Should(Succeed())
```

**With `OnError`:**
```
// Note: we would have to define a wait.Backoff to make sure it similarly to
// Eventually elsewhere.
err := retry.OnError(retry.DefaultRetry, func(_ error) bool {
	// Technically we don't care.
	return true
}, func() error {
	return k8sClient.Get(ctx, helloWorldTFKey, &recheckHelloWorldTF)
})
g.Expect(err).To(Succeed())
```

If I simplify everything, it's still much more complex for no benefits.
```go
err := retry.OnError(retry.DefaultRetry, func(_ error) bool { return true }, func() error {
	return k8sClient.Get(ctx, helloWorldTFKey, &recheckHelloWorldTF)
})
g.Expect(err).To(Succeed())
```

Sure, we can separate the retry logic from the actual call, but
1. We can check that in Eventually too but on in a fancy two-separate-function way.
2. Most of the time we don't care what went wrong, we just want to retry it until it successful or throw an error.

Reply:
An option couldn't be to create a global wrapper around all `Get` and use that in all tests, but than either we create a new package or copy paste the wrapper function all over to all modules and it wouldn't help much in most cases where `Eventually` test things and not just a safeguard.

Reply:
I think

```
var recheckHelloWorldTF infrav1.Terraform
g.Expect(reliableGet(ctx, helloWorldTFKey, &recheckHelloWorldTF)).To(Succeed())
```

is easier to read than

```
var recheckHelloWorldTF infrav1.Terraform
g.Eventually(func() error {
	return k8sClient.Get(ctx, helloWorldTFKey, &recheckHelloWorldTF)
}, timeout, interval).Should(Succeed())
```

 -- it's shorter, but also there's no `Eventually` to make me think that there's more going on than there really is.

> It couldn't be an option to create a global wrapper around all Get and use that in all tests

Why not?

Reply:
Proportionality check: I am not fussed about this particular point, if the tests are right.

Reply:
In general I agree with that, but I wouldn't create a function just because we have 3 instances where we would use it, but I'll keep in mind for example when we get to refactor all the tests to remove unnecessary parts and simplify things because we can.

Reply:
> If I replace `time.Sleep(...)` with `reconciler.Reconcile(...)`, then the test fails; it's _relying_ on not being requeued. But a requeue can happen without being triggered by an update; e.g., every [cache.SyncPeriod](https://github.com/kubernetes-sigs/controller-runtime/blob/v0.16.3/pkg/cache/cache.go#L166).
> 
> To avoid retrying without something changing, you may need to check the generation against the `.observedGeneration` field of the `RetryLimitReached` condition*, and make sure that latter field is set.

This is the last thing to address. My diagnosis and my suggestion above are a ways wrong, now I read them back. I suggested that it should check the observed generation _of the Stalled condition_ -- but that's not right, because what matters is if the spec changed since the last time it tried, whatever happened. The code here already checks the (`.metadata.generation != .status.observedGeneration`](https://github.com/weaveworks/tf-controller/blob/870-retry/controllers/tf_controller.go#L300), which is right.

This is how it appears to work, in pseudo-code:

```
if changed since last time: # L299
  reset failure counter
if failure counter >= retry limit: # L441
  set stalled condition and exit
run program
if program failed: # L467
  increment retry counter
  requeue and exit # L493
else:
  reset failure count
requeue and exit
```
This is correct, on the face of it.

In the test case, once the resource has reached the retry limit, the test sleeps for 15s, then checks it hasn't run again. According to the pseudo-code above (and I believe the code too) that will succeed trivially, because once it's reached the limit, it doesn't requeue it.

However, a resource _can_ be queued in this state, e.g., if the client cache is "synced" (where controller-runtime runs everything in the client cache through the reconciler, to flush out any missed changes). This won't happen in the test, because the sync period is ten hours -- but it would happen, more than twice a day, in practice. The controller should behave according to the pseudo-code above, when it does happen.

If I replace the `time.Sleep()` with `reconciler.Reconcile()`, it simulates the resource being queued arbitrarily. The test now **fails**, because the stalled condition is no longer present. That's surprising! ...given the apparent correspondence of the code and the scheme given above.

I added some printlns to where it decides if the object has changed. It turns out that when it is run _after_ hitting the retry limit, `.status.lastAttemptedRevision` is empty; that is compared to the source revision (L299), which is not empty, and the controller decides it's changed and resets the failure count. So, long story short: generally the code fits the pseudo-code above, but there's a bug in how the status is patched when it hits the retry limit.

Reply:
Added the check to the test run it (without other code changes) and indeed `lastAttemptedRevision` was empty.

I know it does not fix the cache sync issue in test, but this should fix the reset without `lastAttemptedRevision`: https://github.com/weaveworks/tf-controller/pull/1061/commits/935bf5bff8f7dffb4d5ada588022c130c9cea55f ?

Reply:
tf_controller.go L451: When `!ShouldRetry()`, `terraform.TerraformReachedLimit()` is called, and that in turn [calls `SetTerraformReadiness`](https://github.com/weaveworks/tf-controller/pull/1061/files#diff-3f0c6c788c08365729a653db804a723d7755bd4ff5016186f527223e4d4c46d9R858) with an empty string for the last attempted revision.

Reply:
:woman_facepalming: good catch.

Reply:
I don't even know why I called `SetTerraformReadiness` there.